### DISCUSSION TITLE
[decorators] Is Babel working with TypeScript to ensure equivalent behavior of decorators?

### DISCUSSION THREAD

**Discussion by trusktr** [babel/babel#15482] [General]
The pain with legacy decorators before stage 3 decorators was testing decorators to ensure they'd work in every permutation of Babel and TypeScript configs:

- Babel loose:true class fields with decorators v1
- Babel loose:false class fields with decorators v1
- Babel loose class fields with decorators v2
- Babel class fields with decorators v2
- TypeScript useDefineForClassFields:true and experimental decorators
- TypeScript useDefineForClassFields:false and experimental decorators

I had to run my decorator tests 6 times, one per permutation of build config.

With Stage 3 decorators, this is at least better, only two permutations:

- Babel decorators (non-loose class fields, we're moving forward, done with loose mode)
- TypeScript 5 decorators, and useDefineForClassFields (moving forward! if you're using useDefineForClassFields:false, it won't be supported, looking for simplicity and alignment with ES specs)

If Babel and TypeScript teams work together to ensure a set of shared conformance tests pass for both implementations, then we don't really need to worry about running more than one test for decorators, just pick one and go.

Are there conformance tests both teams are using?

This may also extend to other compilers like esbuild and SWC.

**Comment by nicolo-ribaudo** [comment]
Babel has always implemented the current-at-the-time version of the decorators proposal, and TypeScript just released it.

As with other ECMAScript features, we run the test262 test suite: https://github.com/tc39/test262/. TypeScript doesn't run it, however for every other ECMAScript feature it has not been a problem: we both follow the same spec (ECMA-262 + the current decorators proposal), and any divergence from the spec for either Babel or TypeScript is a bug.
